modules/cas_cache/layer_cache_management.c

```diff
@@ -15,141 +15,378 @@ extern u32 seq_cut_off_mb;
+struct _cache_mngt_async_context {
+	struct completion compl;
+	atomic_t ref_count;
+	int result;
+};
+
+struct _cache_mngt_async_lock_context {
+	struct completion compl;
+	spinlock_t lock_sync;
+	int result;
+};
+
+/*
+ * Value used to mark async call as completed. Used when OCF call already
+ * finished, but complete function still has to be performed.
+ */
+#define ASYNC_CALL_FINISHED 1
+
 static void _cache_mngt_lock_complete(ocf_cache_t cache, void *priv, int error)
 {
-	struct _cache_mngt_sync_context *context = priv;
+	struct _cache_mngt_async_lock_context *context = priv;
+	bool interrupted;
+	unsigned long lock_flags = 0;
+
+	spin_lock_irqsave(&context->lock_sync, lock_flags);
+	interrupted = (context->result == -KCAS_ERR_WAITING_INTERRUPTED);
+	if (!error && interrupted)
+		ocf_mngt_cache_unlock(cache);
+	else
+		context->result = error ? error : ASYNC_CALL_FINISHED;
 
-	*context->result = error;
 	complete(&context->compl);
+	spin_unlock_irqrestore(&context->lock_sync, lock_flags);
+	if (interrupted)
+		kfree(context);
 }
 
 static int _cache_mngt_lock_sync(ocf_cache_t cache)
 {
-	struct _cache_mngt_sync_context context;
+	struct _cache_mngt_async_lock_context *context;
 	int result;
 
-	init_completion(&context.compl);
-	context.result = &result;
+	context = kmalloc(sizeof(*context), GFP_KERNEL);
+	if (!context)
+		return -ENOMEM;
 
-	ocf_mngt_cache_lock(cache, _cache_mngt_lock_complete, &context);
-	wait_for_completion_interruptible(&context.compl);
+	init_completion(&context->compl);
+	spin_lock_init(&context->lock_sync);
+
+	ocf_mngt_cache_lock(cache, _cache_mngt_lock_complete, context);
+	result = wait_for_completion_interruptible(&context->compl);
+
+	spin_lock(&context->lock_sync);
+	if (context->result)
+		result = (context->result != ASYNC_CALL_FINISHED) ? context->result : 0;
+	else if (result < 0)
+		result = context->result = -KCAS_ERR_WAITING_INTERRUPTED;
+	spin_unlock(&context->lock_sync);
+	if (result == -KCAS_ERR_WAITING_INTERRUPTED)
+		kfree(context);
 
 	return result;
 }
 
 static int _cache_mngt_read_lock_sync(ocf_cache_t cache)
 {
-	struct _cache_mngt_sync_context context;
+	struct _cache_mngt_async_lock_context *context;
 	int result;
 
-	init_completion(&context.compl);
-	context.result = &result;
+	context = kmalloc(sizeof(*context), GFP_KERNEL);
+	if (!context)
+		return -ENOMEM;
+
+	init_completion(&context->compl);
+	spin_lock_init(&context->lock_sync);
 
-	ocf_mngt_cache_read_lock(cache, _cache_mngt_lock_complete, &context);
-	wait_for_completion_interruptible(&context.compl);
+	ocf_mngt_cache_read_lock(cache, _cache_mngt_lock_complete, context);
+	result = wait_for_completion_interruptible(&context->compl);
+
+	spin_lock(&context->lock_sync);
+	if (context->result)
+		result = (context->result != ASYNC_CALL_FINISHED) ? context->result : 0;
+	else if (result < 0)
+		result = context->result = -KCAS_ERR_WAITING_INTERRUPTED;
+	spin_unlock(&context->lock_sync);
+
+	if (result == -KCAS_ERR_WAITING_INTERRUPTED)
```

Comment:
IOCTL handlers should have reversed condition for freeing context. This comment applies to all IOCTL handlers with this kind of logic (start/stop/lock/(?))